File: src/main/kotlin/com/natu/ftax/ledger/infrastructure/LedgerBookRepositoryJdbc.java
package com.natu.ftax.ledger.infrastructure;

import com.natu.ftax.ledger.domain.LedgerBook;
import jakarta.transaction.Transactional;
import org.springframework.stereotype.Repository;

import java.util.List;

@Repository
public class LedgerBookRepositoryJdbc implements LedgerBookRepository {

    private final LedgerBookRepositoryJpa ledgerBookRepositoryJpa;
    private final LedgerEntryRepository ledgerEntryRepository;
    private final BalanceRepository balanceRepository;

    public LedgerBookRepositoryJdbc(LedgerBookRepositoryJpa ledgerBookRepositoryJpa, LedgerEntryRepository ledgerEntryRepository, BalanceRepository balanceRepository) {
        this.ledgerBookRepositoryJpa = ledgerBookRepositoryJpa;
        this.ledgerEntryRepository = ledgerEntryRepository;
        this.balanceRepository = balanceRepository;
    }

    @Transactional
    @Override
    public void save(LedgerBook ledgerBook) {
        ledgerBook.getLedgerEntries().forEach(entry -> ledgerEntryRepository.save(entry, ledgerBook.getId()));
        balanceRepository.cleanOrphanBalances();
    }

    @Override
    public LedgerBook get() {
        List<LedgerBook> result = ledgerBookRepositoryJpa.findAll().stream().map(LedgerBookEntity::toDomain).toList();
        return result.isEmpty() ? null : result.get(0);
    }

    @Override
    public void delete(String ledgerBookId) {
        ledgerBookRepositoryJpa.deleteById(ledgerBookId);
    }
}

File: src/main/kotlin/com/natu/ftax/ledger/infrastructure/LedgerBookRepositoryJdbc.kt
package com.natu.ftax.ledger.infrastructure

import LedgerBookRepository
import com.natu.ftax.ledger.domain.LedgerBook
import jakarta.transaction.Transactional
import org.springframework.stereotype.Repository


@Repository
class LedgerBookRepositoryJdbc(
    val ledgerBookRepositoryJpa: LedgerBookRepositoryJpa,
    val ledgerEntryRepository: LedgerEntryRepository,
    val balanceRepository : BalanceRepository
) : LedgerBookRepository {
    @Transactional
    override fun save(ledgerBook: LedgerBook) {
        ledgerBook.ledgerEntries.forEach {
            ledgerEntryRepository.save(it, ledgerBook.id)
        }

        balanceRepository.cleanOrphanBalances()
    }

    override fun get(): LedgerBook? {
        val result =  ledgerBookRepositoryJpa.findAll().map { it.toDomain() }.toMutableList()
        return if (result.isEmpty()) null else result[0]
    }

    @Transactional
    override fun delete(ledgerBookId: String) {
        ledgerBookRepositoryJpa.deleteById(ledgerBookId)
        balanceRepository.cleanOrphanBalances()
    }
}